pkg/ocmagenthandler/ocmagenthandler_deployment.go

```diff
@@ -0,0 +1,243 @@
+
+	ocmagentv1alpha1 "github.com/openshift/ocm-agent-operator/pkg/apis/ocmagent/v1alpha1"
+	oahconst "github.com/openshift/ocm-agent-operator/pkg/consts/ocmagenthandler"
+)
+
+func buildOCMAgentDeployment(ocmAgent ocmagentv1alpha1.OcmAgent) appsv1.Deployment {
+	namespacedName := oahconst.BuildNamespacedName()
+	labels := map[string]string{
+		"app": oahconst.OCMAgentName,
+	}
+	labelSelectors := metav1.LabelSelector{
+		MatchLabels: labels,
+	}
+
+	var volumes []corev1.Volume
+	var volumeMounts []corev1.VolumeMount
+	// Define a volume/volume mount for the access token secret
+	var secretVolumeSourceDefaultMode int32 = 0600
+	tokenSecretVolumeName := ocmAgent.Spec.TokenSecret
+	volumes = append(volumes, corev1.Volume{
+		Name: tokenSecretVolumeName,
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName:  ocmAgent.Spec.TokenSecret,
+				DefaultMode: &secretVolumeSourceDefaultMode,
+			},
+		},
+	})
+	volumeMounts = append(volumeMounts, corev1.VolumeMount{
+		Name:      tokenSecretVolumeName,
+		MountPath: filepath.Join(oahconst.OCMAgentSecretMountPath, tokenSecretVolumeName),
+	})
+
+	// Define a volume/volume mount for the config
+	configVolumeName := ocmAgent.Spec.OcmAgentConfig
+	var configVolumeSourceDefaultMode int32 = 0600
+	volumes = append(volumes, corev1.Volume{
+		Name: configVolumeName,
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: ocmAgent.Spec.OcmAgentConfig,
+				},
+				DefaultMode: &configVolumeSourceDefaultMode,
+			},
+		},
+	})
+	volumeMounts = append(volumeMounts, corev1.VolumeMount{
+		Name:      configVolumeName,
+		MountPath: filepath.Join(oahconst.OCMAgentConfigMountPath, configVolumeName),
+	})
+	// Sort volume slices by name to keep the sequence stable.
+	sort.Slice(volumes, func(i, j int) bool {
+		return volumes[i].Name < volumes[j].Name
+	})
+	sort.Slice(volumeMounts, func(i, j int) bool {
+		return volumeMounts[i].Name < volumeMounts[j].Name
+	})
+
+	replicas := int32(ocmAgent.Spec.Replicas)
+	dep := appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      namespacedName.Name,
+			Namespace: namespacedName.Namespace,
+			Labels:    labels,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &labelSelectors,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: labels,
+				},
+				Spec: corev1.PodSpec{
+					Volumes:            volumes,
+					ServiceAccountName: oahconst.OCMAgentServiceAccount,
+					Affinity: &corev1.Affinity{
+						NodeAffinity: &corev1.NodeAffinity{
+							PreferredDuringSchedulingIgnoredDuringExecution: []corev1.PreferredSchedulingTerm{{
+								Preference: corev1.NodeSelectorTerm{
+									MatchExpressions: []corev1.NodeSelectorRequirement{{
+										Key:      "node-role.kubernetes.io/infra",
+										Operator: corev1.NodeSelectorOpExists,
```

Comment:
Considering what we have for MUO and managed-cluster-validating webhooks, what about having this workload on the control plane as well? Intent being, OAO operations do not get hampered/interrupted by an customer actions on worker or infra nodes (saying worker as the affinity is "preferred" here and not "required").

Reply:
We discussed this in 24/1 Hulk Sync and the majority opinion was to leave off the control plane for now, but this is certainly something we can revisit later if we need to (as we did for MUO).